File: Algorithmic_Toolbox_C1/W2_Warm_up/fib_sum_last_num.py
# python3
# Compute the last digit of the sum of n Fibonacci numbers.

import numpy as np
import sys
import time

# Naive implementation, stores whole fibonnaci number, used for stress testing more efficient implementations.
def get_fib_sum_last_digit_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1
    sum = previous + current

    for _ in range(n - 1):
        previous, current = current, (previous + current) % 10
        sum = (sum + current) % 10

    return sum

# More efficient implementation, stores only the ones place of each fibonacci number.
def get_fib_sum_last_digit_table(n):
    fib_table = [0,1]
    fib_sum = [0, 1]
    if n<=1:
        return n

    for i in range(2,n+1):
        fib_table.append((fib_table[i-1] + fib_table[i-2]))
        fib_sum.append(sum(fib_table)%10)

    return fib_sum

# Efficient implementation given that sum of fibonacci numbers mod 10 is a periodic sequence.
def get_fib_sum_last_digit_eff(n):
    fib_sum_table = [0, 1]
    if n <= 1:
        return n

    for index in range(2, n+1):
        fib_sum_table.append((1 + fib_sum_table[index - 1] + fib_sum_table[index - 2]) % 10)
        if (fib_sum_table[index - 1] == 0 and fib_sum_table[index] == 1):
            return fib_sum_table[n % (index - 1)]
    return fib_sum_table[(index)]


def stress_test(fib_range=20):
    counter = 0
    while(True):
        n = np.random.randint(fib_range+1)

        sol1 = get_fib_sum_last_digit_naive(n)
        sol2 = get_fib_sum_last_digit_eff(n)
        assert(sol1 == sol2), f"ERROR: n = {n}\n SOL1: {sol1}\nSOL2: {sol2}"
        print(f"Test {counter} succesful")
        counter += 1

if __name__ == '__main__':

    if len(sys.argv) > 1:
        stress_test(int(sys.argv[1]))
    else:
        n = int(input())
        # print(get_fib_sq_sum_last_digit_naive(n))
        # print(get_fib_sq_sum_last_digit_table(n))
        # start = time.time()
        print(get_fib_sum_last_digit_eff(n))
        # print(f"Total time: {time.time()-start}")

File: Algorithmic_Toolbox_C1/W2_Warm_up/fib_sum_last_num.c
/*
 ============================================================================
 Name        : fib_sum_last_num.c
 Author      : Ivan
 Version     :
 Copyright   : Your copyright notice
 Description : Compute nth fibonacci number mod m Ansi-style
 ============================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <time.h>

#define CHAR_BUFFSIZE (int) (1<<20) // 2^20 = 1048576 bytes around 1 mb

int get_fib_sum_last_num_naive(long n) {
	if(n <= 1) {
		return n;
	}
	
	int past = 0;
	int current = 1;
	int temp;
	int sum = past + current;


	for(long i = 2; i <= n; ++i) {
		temp = past;
		past = current;
		current = (current + temp) % 10;
		sum = (sum + current) % 10;
	}

	return sum;
}

char get_fib_sum_last_num_eff(long n){


	char fib_sum[CHAR_BUFFSIZE+1]; fib_sum[0] = 0; fib_sum[1] = 1;
	
	if(n <= 1)
		return (char) n;

	for(long i = 2; i <= n; ++i) {

		fib_sum[i] = (1 + fib_sum[i-1] + fib_sum[i-2]) % 10;

		if(fib_sum[i-1] == 0 && fib_sum[i] == 1)
			return fib_sum[n % (i-1)];

		if(i % CHAR_BUFFSIZE == 0){
			printf("Pisano period is too large or non-existent\n");
			return 0;
		}
	}

	return fib_sum[n];
}


int stress_test(int max_n){

	int counter = 0;
	int n;
	srand(time(0));
    
    while(1){
		n = rand() % max_n; n++;

        int sol1 = get_fib_sum_last_num_naive(n);
        int sol2 = get_fib_sum_last_num_eff(n);

		char *msg;
		asprintf(&msg, "ERROR: %d\n SOL1: %d\nSOL2: %d", n, sol1, sol2);
        
        assert(sol1 == sol2 && msg);
        free(msg); // release the memory allocated by asprintf.

        printf("Test %d successful\n", counter);
        counter = counter + 1;
    }
}

int main(void) {
	int debug = 0;

	if(debug == 1){
		stress_test(5000);
	}
	else{
		long n;
		scanf("%ld", &n);
		// printf("%d\n", get_fib_sum_last_num_naive(n));
		printf("%d\n", get_fib_sum_last_num_eff(n));
	}

	
	return 0;
}
